packages/teleport/src/useUrlFiltering.tsx

```diff
@@ -0,0 +1,176 @@
+import {
+  Label,
+  LabelTag,
+  LabelOption as Option,
+  makeLabelTag,
+} from 'teleport/services/resources';
+
+// PARAM_FILTER is a url query parameter for filters.
+const PARAM_FILTER = '?filter=';
+
+// FILTER_TYPE_LABEL defines a filter type for labels in a url query param.
+const FILTER_TYPE_LABEL = 'label:';
+
+export default function useUrlFiltering<T extends Filterable>(data: Array<T>) {
+  const history = useHistory();
+  const { search, pathname } = useLocation();
+  const selectedLabels = useMemo<Option[]>(() => getLabelsFromUrl(search), [
+    search,
+  ]);
+
+  const filteredData = useMemo(() => filterData(data, selectedLabels), [
+    data,
+    selectedLabels,
+  ]);
+
+  function apply(labels: Option[]) {
+    history.replace(getEncodedUrl(labels, pathname));
+  }
+
+  // toggleLabel removes an existing label from the
+  // selected labels list, else adds new label to list.
+  function toggleLabel(label: Label) {
+    const tag = makeLabelTag(label);
+    let modifiedList = [...selectedLabels];
+    const index = selectedLabels.findIndex(o => o.label === tag);
+
+    if (index > -1) {
+      // remove the label
+      modifiedList.splice(index, 1);
+    } else {
+      modifiedList = [
+        ...selectedLabels,
+        { label: tag, value: tag, obj: label },
+      ];
+    }
+
+    apply(modifiedList);
+  }
+
+  return {
+    selectedLabels,
+    filteredData,
+    apply,
+    toggleLabel,
+  };
+}
+
+// getLabelsFromUrl parses the query param filter values.
+function getLabelsFromUrl(search: string): Option[] {
+  if (!search.startsWith(PARAM_FILTER)) {
+    return [];
+  }
+
+  const query = search.substring(PARAM_FILTER.length);
+  if (!query) {
+    return [];
+  }
+
+  const filters = query.split('+');
+  const options: Option[] = [];
+
+  for (let i = 0; i < filters.length; i++) {
+    const filter = filters[i];
+    if (!filter) {
+      continue;
+    }
+
+    if (filter.startsWith(FILTER_TYPE_LABEL)) {
+      const encodedLabel = filter.substring(FILTER_TYPE_LABEL.length);
+      const [encodedName, encodedValue, more] = encodedLabel.split('=');
+
+      // Abort if more than two values were split (malformed label).
+      if (more) {
+        return [];
+      }
+
+      const label: Label = {
+        name: decodeURIComponent(encodedName ?? ''),
+        value: decodeURIComponent(encodedValue ?? ''),
+      };
+
+      if (!label.name && !label.value) {
+        continue;
+      }
+
+      const tag = makeLabelTag(label);
+      options.push({
+        label: tag,
+        value: tag,
+        obj: label,
+      });
+
+      continue;
+    }
+
+    // Abort if undefined filter types found among query.
+    return [];
+  }
+
+  return options;
+}
+
+// filterData returns new list of data that contains the selected labels.
+// Return type is relaxed here so it can be passed down to components with
+// stricter types.
+function filterData(data: Filterable[] = [], labels: Option[] = []): any[] {
```

Comment:
`useUrlFiltering` has a generic type so we can use it here:
`filterData<T extends Filterable>(data: T[] = [], labels: Option[] = []): T[]`